Improve error handling for custom/duplicate decorators

pmvald opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

pmvald committed:
fix(compiler-cli): forbid custom/duplicate decorator when option `forbidOrphanComponents` is set

The deps tracker which is responsible to track orphan components does not work for classes mutated by custom decorator. Some work needed to make this happen (tracked in b/320536434). As a result, with option `forbidOrphanComponents` being true the deps tracker will falsely report any component as orphan if it or its NgModule have custom/duplicate decorators. So it is unsafe to use this option in the presence of custom/duplicate decorator and we disable it until it is made compatible. Note that applying custom/duplicate decorators to `@Injectable` classes is ok since these classes never make it into the deps tracker. So we excempt them.

pmvald committed:
fix(compiler-cli): consider the case of duplicate Angular decorators in local compilation diagnostics

For cases like this:

```
@Component({...})
@Component({...})
export class SomeComp {
}
```

The `DecoratorHandler.detect` apparantly matches only one of the `@Component` decorator, leaving the other undetected which will be transformed by TS decorator helper and that breaks local compilation runtimes. But the error message only mentioned "custom" decorator, while in this case it is a "duplicate Angular" decorator. The respective error message is updated thus.

pmvald commented:
[TGP link](https://test.corp.google.com/OCL:602241700:BASE:602266577:1706505538221:670fb4bc)

MeftunH commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -1167,7 +1167,7 @@ export class NgCompiler {
         handlers, reflector, this.delegatingPerfRecorder, this.incrementalCompilation,
         this.options.compileNonExportedClasses !== false, compilationMode, dtsTransforms,
-        semanticDepGraphUpdater, this.adapter);
+        semanticDepGraphUpdater, this.adapter, isCore, !!this.options.forbidOrphanComponents);
```
why we did use magic operator in here?

MeftunH commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -714,3 +745,16 @@ function containsErrors(diagnostics: ts.Diagnostic[]|null): boolean {
+ * imported here due to circular deps. This helper is needed temporary though (see the call site).
+ */
+function isAngularInjectableDecorator(decorator: Decorator, isCore: boolean): boolean {
+  if (isCore) {
+    return decorator.name === 'Injectable';
+  } else if (decorator.import !== null && decorator.import.from === '@angular/core') {
+    return decorator.import.name === 'Injectable';
+  }
```
I think this is unclear usage. So, it makes harder to find changes for next versions

devversion replied:
I'm not sure I follow. Can you elaborate please?

MeftunH commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -1167,7 +1167,7 @@ export class NgCompiler {
         handlers, reflector, this.delegatingPerfRecorder, this.incrementalCompilation,
         this.options.compileNonExportedClasses !== false, compilationMode, dtsTransforms,
-        semanticDepGraphUpdater, this.adapter);
+        semanticDepGraphUpdater, this.adapter, isCore, !!this.options.forbidOrphanComponents);
 
```
why we did use magic operator in here?

devversion commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -259,19 +257,25 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
+    // reason for both cases is that the `DepsTracker` is not able to understand components mutated
+    // by decorators. Once the `DepsTracker` is fixed then these issues will be resolved. This issue
+    // is to be fixed soon (b/320536434).
+    const unmatchedDecoratorsInLocalMode =
```
```suggestion
    const unmatchedDecorators =
```

Regarding "fixing deps tracker". There is no fix we are landing right? except hoping for ES decorators at some point. The comment sound like deps tracker needs a fix.

devversion replied:
I was just curious because of the comment sounding like we intend to allow this in the future again. Maybe remove?

devversion commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -341,21 +345,48 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
+        // don't support custom decorators yet. But will eventually do (b/320536434). For now a
+        // temporary error is thrown.
+
+        // We allow duplicate `@Injectable` in full compilation mode since: 1) it is happening in G3
```
What would it take to fix these usages in g3? If it's less than <= 30 then we should just fix them I think

pmvald replied:
What I'm going to do is to comment out this allowing, and run TGP. By tomorrow morning we will see how many targets are broken. I hope we see a few.

pmvald replied:
[TGP link](https://test.corp.google.com/OCL:602241700:BASE:602571353:1706584359213:eee681fc)

devversion replied:
Hmm, doesn't look like a lot. Like 5-6 "apps". I will leave that to your decision, but might be better to just opt them out of local compilation for now?

pmvald replied:
Yeah, few apps. But the number of individual files, and potentially targets, is ~ 100. At the moment we can opt out the "orphan component check" on targets basis. So it's a bit of work to clean of these up. On the other hand all these cases are related to `@Injectable` classes. So I made these changes: we forbid multi decorator for both local compilation and `forbidOrphanComponents=true` for all classes except injectables. The injectables do not affect deps tracker and so they don't screw things up.

devversion replied:
I don't like this inconsistency but it makes sense to move forward. Can you create a TODO + ticket for tracking?

pmvald replied:
Done! Yeah, probably I'll fix it soon. I do have a tentative solution ib mind, but need to write a DD for it.

devversion commented on packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts:
```diff
@@ -40,7 +40,7 @@ runInEachFileSystem(() => {
       const compiler = new TraitCompiler(
           handlers, reflectionHost, NOOP_PERF_RECORDER, NOOP_INCREMENTAL_BUILD, true,
-          compilationMode, new DtsTransformRegistry(), null, fakeSfTypeIdentifier);
+          compilationMode, new DtsTransformRegistry(), null, fakeSfTypeIdentifier, false, false);
```
Nit: add comments what the boolean arguments represent

pmvald committed:
fix(compiler-cli): allow custom/duplicate decorators for @Injectable classes in local compilation mode

Custom/duplicate decorators break the deps tracker in local mode. But deps tracker only deals with non-injectable classes. So applying custom/duplicate decorators to `@Injectable` only classes does not disturb deps tracker and local compilation in general. There are also ~ 100 such cases in g3 which cannot be cleaned up.

devversion commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -714,3 +728,31 @@ function containsErrors(diagnostics: ts.Diagnostic[]|null): boolean {
+ */
+function hasScopeDecorators(decoratorSet: Set<Decorator>, isCore: boolean) {
+  for (const decorator of decoratorSet) {
+    if (!isInjectableDecorator(decorator, isCore)) {
```
Have you considered using `findAngularDecorator(decorators, 'Injectable', this.isCore)`?

pmvald replied:
Yes, but it is in the package `//ngtsc/annotations` which already has dependency on this package. So importing it here will produce cyclic deps. That's why I'm duplicating the logic there into `isInjectableDecorator` here.

devversion commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -714,3 +728,31 @@ function containsErrors(diagnostics: ts.Diagnostic[]|null): boolean {
+ * Scope decorators are those who participate in determining the scope of an NgModule or a
+ * standalone component. They are: `@Component`, `@Directive`, `@Pipe` and `@NgModule`.
+ */
+function hasScopeDecorators(decoratorSet: Set<Decorator>, isCore: boolean) {
```
Super nit (if you are at it): Maybe also `hasDepsTrackerAffectingScopeDecorator`

pmvald replied:
Done!

devversion commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -341,20 +342,33 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
-            messageText:
-                'In local compilation mode, Angular does not support custom decorators. Ensure all class decorators are from Angular.',
-          }));
+    if (decorators !== null && detectedDecorators !== null &&
```
Please add a small comment why we are disallowing above the `if`. I.e. for the deps tracker on which local compilation relies on.

pmvald replied:
Done!

devversion reviewed (approved):
LGTM, please add merge ready once you think it's ready

jessicajaniuk commented:
This PR was merged into the repository by commit 96bcf4fb1208d1f073784a2cde4a03553e905807.

jessicajaniuk commented:
@pmvald This caused a breakage in G3 and has been reverted.

JeanMeche commented:
This PR has gone stale.
